protocol/communities/manager.go

```diff
@@ -1530,6 +1600,157 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+	indexFile.Download()
+	for {
+		if indexFile.BytesCompleted() == indexFile.Length() {
+			break
+		}
+	}
+
+	index, err := m.LoadHistoryArchiveIndexFromFile(communityID)
+	if err != nil {
+		return nil, err
+	}
+
+	var archiveIDs []string
+
+	archiveHashes := make(archiveMDSlice, 0, len(index.Archives))
+
+	for hash, metadata := range index.Archives {
+		archiveHashes = append(archiveHashes, &archiveMetadata{hash: hash, from: metadata.Metadata.From})
+	}
+
+	sort.Sort(archiveHashes)
+
+	for _, hd := range archiveHashes {
+		hash := hd.hash
+		metadata := index.Archives[hash]
+		hasArchive, err := m.persistence.HasMessageArchiveID(communityID, hash)
+		if err != nil {
+			m.logger.Debug("Failed to check if has message archive id", zap.Error(err))
+			continue
+		}
+		if hasArchive {
+			continue
+		}
+
+		startIndex := int(metadata.Offset) / pieceLength
+		endIndex := startIndex + int(metadata.Size)/pieceLength
+
+		m.logger.Debug("Downloading data for message archive", zap.Any("hash", hash))
+		m.logger.Debug("Pieces (start, end)", zap.Any("startIndex", startIndex), zap.Any("endIndex", endIndex-1))
+		torrent.DownloadPieces(startIndex, endIndex)
+
+		psc := torrent.SubscribePieceStateChanges()
+		for {
+			i := startIndex
+			done := false
+			for {
+				if i > endIndex-1 {
+					break
+				}
+				done = torrent.PieceState(i).Complete
+				i++
+			}
+			if done {
+				psc.Close()
+				break
+			}
+			<-psc.Values
+		}
+		archiveIDs = append(archiveIDs, hash)
+		err = m.persistence.SaveMessageArchiveID(communityID, hash)
+		if err != nil {
+			m.logger.Debug("Couldn't save message archive ID", zap.Error(err))
+			continue
+		}
+		m.publish(&Subscription{
+			HistoryArchiveDownloadedSignal: &signal.HistoryArchiveDownloadedSignal{
+				CommunityID: communityID.String(),
+				From:        int(metadata.Metadata.From),
+				To:          int(metadata.Metadata.To),
+			},
+		})
+	}
+	m.publish(&Subscription{
+		HistoryArchivesSeedingSignal: &signal.HistoryArchivesSeedingSignal{
+			CommunityID: communityID.String(),
+		},
+	})
+	return archiveIDs, nil
+}
+
+func (m *Manager) ExtractMessagesFromHistoryArchives(communityID types.HexBytes, archiveIDs []string) (map[transport.Filter][]*types.Message, error) {
+	id := communityID.String()
+
+	index, err := m.LoadHistoryArchiveIndexFromFile(communityID)
+	if err != nil {
+		return nil, err
+	}
+	totalData, err := os.ReadFile(m.archiveDataFile(id))
+	if err != nil {
+		return nil, err
+	}
+
+	messages := make(map[transport.Filter][]*types.Message)
+
+	for _, hash := range archiveIDs {
+		metadata := index.Archives[hash]
+
+		archive := &protobuf.WakuMessageArchive{}
+		data := totalData[metadata.Offset : metadata.Offset+metadata.Size-metadata.Padding]
+
+		err := proto.Unmarshal(data, archive)
+		if err != nil {
+			m.logger.Debug("Failed to unmarshal WakuMessageArchive", zap.Error(err))
+			continue
+		}
+
+		for _, message := range archive.Messages {
+			filter := m.transport.FilterByTopic(message.Topic)
```

Comment:
what are you checking exactly here? meaning, you are checking the topic matches, but generally that's not enough to ensure the message is for you and you should process it. 
Which messages should we process and in which case they should not be processed?

Reply:
This function (`ExtractMessagesFromHistoryArchive`) aims to create a `map[transport.Filter][]*types.Message` from a downloaded archive, so they can be passed to `handleRetrievedMessages`.

Because we're only getting the topic from each archive message, we need to figure out what filter they belong to, so we can create the correct return type.
All messages extracted from an archive (and passed to `handleRetrievedMessages`) are considered to be processed.